src/core/Object.h

```diff
@@ -29,39 +29,119 @@
 
 #include <unistd.h>
 #include <iostream>
+#include <atomic>
+#include <map>
 #include <QtCore>
 #include <QDebug>
 
 namespace H2Core {
 
+/** an objects class map item type */
+typedef struct atomic_obj_cpt_t {
+	std::atomic<int> constructed;
+	std::atomic<int> destructed;
+	atomic_obj_cpt_t() : constructed(0), destructed(0) {}
+} atomic_obj_cpt_t;
+typedef struct {
+	int constructed;
+	int destructed;
+} obj_cpt_t;
+
+
+/** the objects class map types */
+typedef std::map<const char*, obj_cpt_t> object_map_t;
+typedef std::map<const char*, const atomic_obj_cpt_t*> object_internal_map_t;
+
+class CountControl {
+	public:
+		static void set_count( bool flag );
+		static bool count_active() { return __count; }
+		static int objects_count() { return __objects_count; }
+		CountControl() {
+#ifdef H2CORE_HAVE_DEBUG
+			if ( __count ) {
+				++__objects_count;
+			}
+#endif
+		}
+		CountControl(const CountControl &other) {
+#ifdef H2CORE_HAVE_DEBUG
+			if ( __count ) {
+				++__objects_count;
+			}
+#endif
+		}
+	protected:
+		~CountControl() {
+#ifdef H2CORE_HAVE_DEBUG
+			if ( __count ) {
+				--__objects_count;
+			}
+#endif
+		}
+		static bool __count;               ///< should we count class instances
+		static std::atomic<int> __objects_count;        ///< total objects count
+		static Logger * __logger;
+};
+
+template<typename T> class Countable: public CountControl {
+	public:
+		Countable() {
+#ifdef H2CORE_HAVE_DEBUG
+			if ( __logger && __logger->should_log( Logger::Constructors ) ) {
+				__logger->log( Logger::Debug, nullptr, T::_class_name(), "Constructor" );
+			}
+			if ( __count ) {
+				++counters.constructed;
+			}
+#endif
+		}
+		Countable(const Countable<T> &other) {
+#ifdef H2CORE_HAVE_DEBUG
+			if( __logger && __logger->should_log( Logger::Constructors ) ) {
+				__logger->log( Logger::Debug, nullptr, T::_class_name(), "Copy Constructor" );
+			}
+			if ( __count ) {
+				++counters.destructed;
```

Comment:
This one is a plain typo, it should read ++counters.constructed, of course. Strange that I didn't caught this one since I just restored the constructor/destructor counting... Hum, maybe that's could mean I introduced the first actual copy operation with my work on envelope points... I'll rebase this one to see what we get.